File: src/foundry/connectors/comments.py
"""Rendering Foundry's view back into the tracker.

Two things the orchestrator writes to Linear:

- a concise analysis/plan comment when a run is planned, and
- a ``Foundry: ...`` workflow state that mirrors the run status.

Comments are kept short and skimmable - the ticket is the system of record for
delivery status, not a wall of text.
"""

from __future__ import annotations

from foundry.schemas.analysis import TicketAnalysis
from foundry.schemas.common import RunStatus
from foundry.schemas.plan import DeliveryPlan
from foundry.schemas.risk import RiskAssessment

# Suggested Linear workflow states, mapped from the run status.
_STATE_NAMES: dict[RunStatus, str] = {
    RunStatus.ANALYSING: "Foundry: Analysing",
    RunStatus.NEEDS_CLARIFICATION: "Foundry: Needs Clarification",
    RunStatus.PLAN_READY: "Foundry: Plan Ready",
    RunStatus.WAITING_APPROVAL: "Foundry: Waiting Approval",
    RunStatus.APPROVED: "Foundry: Approved",
    RunStatus.AGENT_RUNNING: "Foundry: Agent Running",
    RunStatus.PR_OPEN: "Foundry: PR Open",
    RunStatus.REVIEW_REQUIRED: "Foundry: Review Required",
    RunStatus.COMPLETE: "Foundry: Complete",
    RunStatus.BLOCKED: "Foundry: Blocked",
    RunStatus.EXECUTION_FAILED: "Foundry: Blocked",
    RunStatus.REJECTED: "Foundry: Blocked",
}


def state_for(status: RunStatus) -> str:
    return _STATE_NAMES.get(status, f"Foundry: {status.value.replace('_', ' ').title()}")


def _bullets(items: list[str], *, empty: str = "_none_") -> str:
    return "\n".join(f"- {i}" for i in items) if items else empty


def draft_acceptance_criteria(analysis: TicketAnalysis) -> list[str]:
    """Draft acceptance-criteria skeletons the ticket author can edit.

    Clarification should do work for the author, not just bounce the ticket:
    a copy-editable draft converts "needs clarification" from a rejection into
    a 30-second fix.
    """
    drafts = [
        f"Given <starting state>, when {analysis.title.strip().rstrip('.').lower()}, "
        "then <observable outcome>",
    ]
    if "reproduction steps" in analysis.missing_information:
        drafts.append(
            "Steps to reproduce: 1. <go to...> 2. <do...> 3. <observe...>"
        )
    if "a description of the desired outcome" in analysis.missing_information:
        drafts.append("The desired outcome is <what the user should see/be able to do>")
    drafts.append("Out of scope: <anything this ticket deliberately does not change>")
    return drafts


def format_analysis_comment(
    analysis: TicketAnalysis,
    risk: RiskAssessment,
    plan: DeliveryPlan,
    status: RunStatus,
    *,
    required_roles: list[str] | None = None,
) -> str:
    """Render the planning summary comment posted to the issue.

    ``required_roles`` is the *effective* set of approval roles a human must
    hold - the risk-derived roles unioned with any per-repo roles scoped to the
    routed repo (issue #31). It defaults to the risk-derived roles so callers
    that don't compute the union still render correctly.
    """
    repo = plan.affected_repositories[0] if plan.affected_repositories else "_unknown_"
    approvals = (
        list(required_roles)
        if required_roles is not None
        else [r.value for r in risk.required_approvals]
    )
    lines = [
        "**Foundry analysis complete.**",
        "",
        f"- Work type: `{analysis.work_type.value}`",
        f"- Readiness: `{analysis.implementation_readiness.value}`",
        f"- Risk: `{risk.overall_risk.value}` "
        f"(suggested mode: `{risk.allowed_agent_mode.value}`)",
        f"- Affected repo: `{repo}`",
        "",
        "**Acceptance criteria**",
        _bullets(analysis.acceptance_criteria),
    ]
    if analysis.missing_information:
        lines += ["", "**Missing information**", _bullets(analysis.missing_information)]
    if plan.implementation_steps:
        lines += [
            "",
            "**Plan**",
            "\n".join(f"{s.step}. {s.description}" for s in plan.implementation_steps),
        ]
    if approvals:
        lines += ["", f"**Required approval:** {', '.join(approvals)}"]

    if status is RunStatus.WAITING_APPROVAL:
        lines += [
            "",
            "Reply to proceed:",
            "`/foundry approve` · `/foundry reject` · `/foundry stop`",
        ]
    elif status is RunStatus.NEEDS_CLARIFICATION:
        lines += [
            "",
            "_Needs clarification before an agent can start._",
            "",
            "**Suggested acceptance criteria** (edit these and add them to the "
            "ticket, then re-trigger Foundry):",
            _bullets(draft_acceptance_criteria(analysis)),
        ]
    elif status is RunStatus.BLOCKED:
        lines += ["", "_Blocked: " + "; ".join(risk.risk_reasons or ["see policy"]) + "._"]
    return "\n".join(lines)


def format_cursor_delegation(agent_instructions: str) -> str:
    """The @Cursor delegation comment that hands approved work to Cursor.

    Foundry has already gathered context, classified risk and obtained approval;
    this passes the *governed* instructions to Cursor's Linear integration, which
    runs the cloud agent, reports status in Linear and opens the PR.
    """
    return (
        "@Cursor please implement this. Work strictly within the scope below; "
        "Foundry has approved it.\n\n"
        f"{agent_instructions}"
    )

File: dotnet/src/Foundry/Connectors/Comments.cs
// Rendering Foundry's view back into the tracker.
//
// Two things the orchestrator writes to Linear:
//
// - a concise analysis/plan comment when a run is planned, and
// - a "Foundry: ..." workflow state that mirrors the run status.
//
// Comments are kept short and skimmable - the ticket is the system of record
// for delivery status, not a wall of text.

using Foundry.Schemas;

namespace Foundry.Connectors;

public static class Comments
{
    // Suggested Linear workflow states, mapped from the run status.
    private static readonly IReadOnlyDictionary<RunStatus, string> StateNames =
        new Dictionary<RunStatus, string>
        {
            [RunStatus.Analysing] = "Foundry: Analysing",
            [RunStatus.NeedsClarification] = "Foundry: Needs Clarification",
            [RunStatus.PlanReady] = "Foundry: Plan Ready",
            [RunStatus.WaitingApproval] = "Foundry: Waiting Approval",
            [RunStatus.Approved] = "Foundry: Approved",
            [RunStatus.AgentRunning] = "Foundry: Agent Running",
            [RunStatus.PrOpen] = "Foundry: PR Open",
            [RunStatus.ReviewRequired] = "Foundry: Review Required",
            [RunStatus.Complete] = "Foundry: Complete",
            [RunStatus.Blocked] = "Foundry: Blocked",
            [RunStatus.ExecutionFailed] = "Foundry: Blocked",
            [RunStatus.Rejected] = "Foundry: Blocked",
        };

    public static string StateFor(RunStatus status) =>
        StateNames.TryGetValue(status, out var name)
            ? name
            : $"Foundry: {Wire.ToTitle(status.ToWire())}";

    private static string Bullets(IReadOnlyList<string> items, string empty = "_none_") =>
        items.Count > 0 ? string.Join("\n", items.Select(i => $"- {i}")) : empty;

    /// <summary>
    /// Draft acceptance-criteria skeletons the ticket author can edit.
    ///
    /// Clarification should do work for the author, not just bounce the
    /// ticket: a copy-editable draft converts "needs clarification" from a
    /// rejection into a 30-second fix.
    /// </summary>
    public static List<string> DraftAcceptanceCriteria(TicketAnalysis analysis)
    {
        var drafts = new List<string>
        {
            $"Given <starting state>, when {analysis.Title.Trim().TrimEnd('.').ToLowerInvariant()}, "
            + "then <observable outcome>",
        };
        if (analysis.MissingInformation.Contains("reproduction steps"))
        {
            drafts.Add("Steps to reproduce: 1. <go to...> 2. <do...> 3. <observe...>");
        }
        if (analysis.MissingInformation.Contains("a description of the desired outcome"))
        {
            drafts.Add("The desired outcome is <what the user should see/be able to do>");
        }
        drafts.Add("Out of scope: <anything this ticket deliberately does not change>");
        return drafts;
    }

    /// <summary>Render the planning summary comment posted to the issue.</summary>
    public static string FormatAnalysisComment(
        TicketAnalysis analysis, RiskAssessment risk, DeliveryPlan plan, RunStatus status)
    {
        var repo = plan.AffectedRepositories.Count > 0 ? plan.AffectedRepositories[0] : "_unknown_";
        var approvals = risk.RequiredApprovals.Select(r => r.ToWire()).ToList();
        var lines = new List<string>
        {
            "**Foundry analysis complete.**",
            "",
            $"- Work type: `{analysis.WorkType.ToWire()}`",
            $"- Readiness: `{analysis.ImplementationReadiness.ToWire()}`",
            $"- Risk: `{risk.OverallRisk.ToWire()}` (suggested mode: `{risk.AllowedAgentMode.ToWire()}`)",
            $"- Affected repo: `{repo}`",
            "",
            "**Acceptance criteria**",
            Bullets(analysis.AcceptanceCriteria),
        };
        if (analysis.MissingInformation.Count > 0)
        {
            lines.AddRange(new[] { "", "**Missing information**", Bullets(analysis.MissingInformation) });
        }
        if (plan.ImplementationSteps.Count > 0)
        {
            lines.AddRange(new[]
            {
                "",
                "**Plan**",
                string.Join("\n", plan.ImplementationSteps.Select(s => $"{s.Step}. {s.Description}")),
            });
        }
        if (approvals.Count > 0)
        {
            lines.AddRange(new[] { "", $"**Required approval:** {string.Join(", ", approvals)}" });
        }

        if (status == RunStatus.WaitingApproval)
        {
            lines.AddRange(new[]
            {
                "",
                "Reply to proceed:",
                "`/foundry approve` · `/foundry reject` · `/foundry stop`",
            });
        }
        else if (status == RunStatus.NeedsClarification)
        {
            lines.AddRange(new[]
            {
                "",
                "_Needs clarification before an agent can start._",
                "",
                "**Suggested acceptance criteria** (edit these and add them to the "
                + "ticket, then re-trigger Foundry):",
                Bullets(DraftAcceptanceCriteria(analysis)),
            });
        }
        else if (status == RunStatus.Blocked)
        {
            var blockedReasons = risk.RiskReasons.Count > 0
                ? risk.RiskReasons
                : new[] { "see policy" };
            lines.AddRange(new[] { "", "_Blocked: " + string.Join("; ", blockedReasons) + "._" });
        }
        return string.Join("\n", lines);
    }

    /// <summary>
    /// The @Cursor delegation comment that hands approved work to Cursor.
    ///
    /// Foundry has already gathered context, classified risk and obtained
    /// approval; this passes the *governed* instructions to Cursor's Linear
    /// integration, which runs the cloud agent, reports status in Linear and
    /// opens the PR.
    /// </summary>
    public static string FormatCursorDelegation(string agentInstructions) =>
        "@Cursor please implement this. Work strictly within the scope below; "
        + "Foundry has approved it.\n\n"
        + agentInstructions;
}
